### PR TITLE
Enable more Ruff checkers and fixers

### PR DESCRIPTION
**Pull request by tiran**
Ruff can not just detect and report code violations. It can automatically fix a large amount of code violations, too. This change enables most checkers and addresses their code violations:

- Update minimal Python version to 3.10
- Remove unnecesary `zip`
- Remove some unused imports
- Simplify some code
- Avoid `l`, which can be easily confused with `1`.
- Use full path regexps for pylint's `ignore-paths`

This change only contains uncontroversial and simple fixes. I will enable more fixers in the near future.

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the
  [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary).
- [ ] [Changelog](https://github.com/instructlab/instructlab/blob/main/CHANGELOG.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

File: src/instructlab/llamacpp/llamacpp_convert_to_gguf.py [hunk 1]
```diff
@@ -48,6 +48,9 @@
     # Standard
     from typing import TypeAlias
 
+# ruff: noqa: SIM115
```
Comment (booxter) [inline src/instructlab/llamacpp/llamacpp_convert_to_gguf.py:51]: I was told elsewhere (by @bjhargrave) that this file (and other files that originate from other repos) should not be "fixed" by linters and e.g. mypy checker. (I disagree btw, but that's beyond the point.) If that's so, this change should probably also *not* happen. (mlx_explore and lora_mlx were also mentioned, I believe.)

Comment (bjhargrave) [reply src/instructlab/llamacpp/llamacpp_convert_to_gguf.py:51]: I am just conscious about excess modifications to 3rd party source we copied in. Sure, add features and fix things that break us. But modifying the code to pass our views on code linting seems a bit over the top and makes it harder to merge in a change from upstream if we want.

Comment (tiran) [reply src/instructlab/llamacpp/llamacpp_convert_to_gguf.py:51]: Makes sense. I'm going to exclude the files from Ruff for now.

We should consider to make vendored files into a `instructlab.vendor` package, so it is obvious that the files were vendored from 3rd party projects.

**Comment by booxter** [comment]
lgtm

**Review by bjhargrave — APPROVED** [review]

**Review by leseb — APPROVED** [review]

**Pull request merged**